File: src/backend/game-logic/tic-tac-toe.js
// src/backend/game-logic/tic-tac-toe.js
const GameEngineInterface = require('./game-engine-interface');

class TicTacToeEngine extends GameEngineInterface {
    constructor() {
        super();
        this.engine_id = 'tic-tac-toe';
        this.winningLines = [
            [0, 1, 2], [3, 4, 5], [6, 7, 8], // rows
            [0, 3, 6], [1, 4, 7], [2, 5, 8], // columns
            [0, 4, 8], [2, 4, 6] // diagonals
        ];
    }

    getEngineId() {
        return this.engine_id;
    }

    initializeGame(options = {}) {
        const playerStarts = options.playerStarts !== false; // Default true
        const difficulty = options.difficulty || 'medium';

        return {
            gameId: this.engine_id,
            board: Array(9).fill(null),
            currentPlayer: playerStarts ? 'X' : 'O',
            gameOver: false,
            winner: null,
            moveHistory: [],
            playerStarts,
            difficulty,
            moveCount: 0,
            createdAt: new Date().toISOString()
        };
    }

    isValidMove(gameState, move) {
        const movePosition = move.position;

        // Check if game is over
        if (gameState.gameOver) {
            return false;
        }

        // Check if position is valid
        if (movePosition < 0 || movePosition > 8) {
            return false;
        }

        // Check if position is empty
        return gameState.board[movePosition] === null;
    }

    processMove(gameState, move) {
        if (!this.isValidMove(gameState, move)) {
            throw new Error('Invalid move');
        }

        const newState = JSON.parse(JSON.stringify(gameState));
        const movePosition = move.position;

        // Make the move
        newState.board[movePosition] = newState.currentPlayer;
        newState.moveCount++;

        // Add to move history
        newState.moveHistory.push({
            player: newState.currentPlayer,
            position: movePosition,
            timestamp: new Date().toISOString(),
            moveNumber: newState.moveCount
        });

        // Check for game end
        const gameEnd = this.checkGameEnd(newState);
        if (gameEnd) {
            newState.gameOver = gameEnd.gameOver;
            newState.winner = gameEnd.winner;
        } else {
            // Switch players
            newState.currentPlayer = newState.currentPlayer === 'X' ? 'O' : 'X';
        }

        return newState;
    }

    checkGameEnd(gameState) {
        // Check for winner
        for (let line of this.winningLines) {
            const [a, b, c] = line;
            if (gameState.board[a] &&
                gameState.board[a] === gameState.board[b] &&
                gameState.board[a] === gameState.board[c]) {
                return {
                    gameOver: true,
                    winner: gameState.board[a]
                };
            }
        }

        // Check for tie
        if (gameState.board.every(cell => cell !== null)) {
            return {
                gameOver: true,
                winner: 'tie'
            };
        }

        return null; // Game continues
    }

    serializeState(gameState) {
        const boardState = gameState.board.map(cell => cell || '_').join('');
        const moveSequence = gameState.moveHistory
            .map(move => `${move.player}:${move.position}`)
            .join(',');

        return {
            boardState,
            metadata: {
                moveSequence,
                moveCount: gameState.moveCount,
                difficulty: gameState.difficulty,
                playerStarts: gameState.playerStarts
            }
        };
    }

    getAIMove(gameState, difficulty = 'medium') {
        if (gameState.currentPlayer !== 'O' || gameState.gameOver) {
            return null;
        }

        const availablePositions = gameState.board
            .map((cell, index) => cell === null ? index : null)
            .filter(pos => pos !== null);

        if (availablePositions.length === 0) {
            return null;
        }

        switch (difficulty) {
            case 'easy':
                return { position: this._getRandomMove(availablePositions) };
            case 'medium':
                return { position: this._getMediumMove(gameState, availablePositions) };
            case 'hard':
                return { position: this._getBestMove(gameState, availablePositions) };
            default:
                return { position: this._getMediumMove(gameState, availablePositions) };
        }
    }

    _getRandomMove(availablePositions) {
        return availablePositions[Math.floor(Math.random() * availablePositions.length)];
    }

    _getMediumMove(gameState, availablePositions) {
        // Try to win first
        const winMove = this._findWinningMove(gameState, 'O');
        if (winMove !== null) return winMove;

        // Try to block player win
        const blockMove = this._findWinningMove(gameState, 'X');
        if (blockMove !== null) return blockMove;

        // Take center if available
        if (availablePositions.includes(4)) return 4;

        // Take corners
        const corners = [0, 2, 6, 8].filter(pos => availablePositions.includes(pos));
        if (corners.length > 0) {
            return corners[Math.floor(Math.random() * corners.length)];
        }

        // Take any remaining position
        return this._getRandomMove(availablePositions);
    }

    _getBestMove(gameState, availablePositions) {
        let bestScore = -Infinity;
        let bestMove = availablePositions[0];

        for (let pos of availablePositions) {
            const testBoard = [...gameState.board];
            testBoard[pos] = 'O';
            const score = this._minimax(testBoard, 0, false);
            if (score > bestScore) {
                bestScore = score;
                bestMove = pos;
            }
        }

        return bestMove;
    }

    _findWinningMove(gameState, player) {
        for (let line of this.winningLines) {
            const [a, b, c] = line;
            const positions = [gameState.board[a], gameState.board[b], gameState.board[c]];

            if (positions.filter(p => p === player).length === 2 &&
                positions.filter(p => p === null).length === 1) {
                // Found a winning opportunity
                if (gameState.board[a] === null) return a;
                if (gameState.board[b] === null) return b;
                if (gameState.board[c] === null) return c;
            }
        }
        return null;
    }

    _minimax(board, depth, isMaximizing) {
        const winner = this._checkWinner(board);
        if (winner === 'O') return 1;
        if (winner === 'X') return -1;
        if (board.every(cell => cell !== null)) return 0;

        if (isMaximizing) {
            let bestScore = -Infinity;
            for (let i = 0; i < 9; i++) {
                if (board[i] === null) {
                    board[i] = 'O';
                    const score = this._minimax(board, depth + 1, false);
                    board[i] = null;
                    bestScore = Math.max(score, bestScore);
                }
            }
            return bestScore;
        } else {
            let bestScore = Infinity;
            for (let i = 0; i < 9; i++) {
                if (board[i] === null) {
                    board[i] = 'X';
                    const score = this._minimax(board, depth + 1, true);
                    board[i] = null;
                    bestScore = Math.min(score, bestScore);
                }
            }
            return bestScore;
        }
    }

    _checkWinner(board) {
        for (let line of this.winningLines) {
            const [a, b, c] = line;
            if (board[a] && board[a] === board[b] && board[a] === board[c]) {
                return board[a];
            }
        }
        return null;
    }

    getStatFields() {
        return [
            { name: 'total_games', type: 'count', label: 'Total Games' },
            { name: 'wins', type: 'count', label: 'Wins' },
            { name: 'losses', type: 'count', label: 'Losses' },
            { name: 'ties', type: 'count', label: 'Ties' },
            { name: 'avg_moves', type: 'average', label: 'Average Moves' },
            { name: 'win_rate', type: 'percentage', label: 'Win Rate' },
            { name: 'incomplete_games', type: 'count', label: 'Incomplete Games' }
        ];
    }
}

module.exports = TicTacToeEngine;

File: src/backend/game_logic/tic_tac_toe.py
"""Pure game logic for Tic-Tac-Toe."""
import random
from typing import Any, Dict, List, Optional


class TicTacToe:
    """Tic-Tac-Toe game logic with difficulty-based AI (easy/medium/hard minimax)."""

    WINNING_LINES = [
        [0, 1, 2], [3, 4, 5], [6, 7, 8],
        [0, 3, 6], [1, 4, 7], [2, 5, 8],
        [0, 4, 8], [2, 4, 6],
    ]

    def get_initial_state(
        self, difficulty: str = "medium", player_starts: bool = True
    ) -> Dict[str, Any]:
        """Return the starting board state for a new Tic-Tac-Toe game.

        Args:
            difficulty: AI difficulty level — "easy", "medium", or "hard".
            player_starts: If True, player (X) moves first; otherwise AI (O) moves first.

        Returns:
            dict with keys: board (9-element list), current_player, player_symbol,
            ai_symbol, game_over, winner, difficulty, player_starts.
        """
        return {
            "board": [None] * 9,
            "current_player": "X" if player_starts else "O",
            "player_symbol": "X",
            "ai_symbol": "O",
            "game_over": False,
            "winner": None,
            "difficulty": difficulty,
            "player_starts": player_starts,
        }

    def apply_move(self, game_state: Dict[str, Any], position: int) -> Dict[str, Any]:
        """Apply a player move, then generate and apply the AI response.

        Args:
            game_state: Current game state dict from get_initial_state or a prior apply_move.
            position: Board index (0–8) for the player's move.

        Returns:
            dict with keys: player_move, board_after_player, game_over_after_player,
            ai_move, board_after_ai, game_over, winner.

        Raises:
            ValueError: If the game is already over, position is out of range, or
                the position is already taken.
        """
        board = game_state["board"][:]
        difficulty = game_state["difficulty"]

        if game_state["game_over"]:
            raise ValueError("Game is already over")
        if not (0 <= position <= 8):
            raise ValueError("Invalid position")
        if board[position] is not None:
            raise ValueError("Position already taken")

        board[position] = game_state["player_symbol"]
        player_board = {**game_state, "board": board}

        winner = self._check_winner(board)
        if winner or self._is_board_full(board):
            outcome = winner if winner else "tie"
            player_board = {**player_board, "game_over": True, "winner": outcome}
            return {
                "player_move": position,
                "board_after_player": player_board,
                "game_over_after_player": True,
                "ai_move": None,
                "board_after_ai": None,
                "game_over": True,
                "winner": outcome,
            }

        ai_position = self._get_ai_move(board, difficulty)
        board = board[:]
        board[ai_position] = game_state["ai_symbol"]

        winner = self._check_winner(board)
        game_over = bool(winner or self._is_board_full(board))
        outcome = winner if winner else ("tie" if game_over else None)
        ai_board = {
            **player_board,
            "board": board,
            "game_over": game_over,
            "winner": outcome,
        }

        return {
            "player_move": position,
            "board_after_player": player_board,
            "game_over_after_player": False,
            "ai_move": ai_position,
            "board_after_ai": ai_board,
            "game_over": game_over,
            "winner": outcome,
        }

    def _get_ai_move(self, board: List[Optional[str]], difficulty: str) -> int:
        available = [i for i, cell in enumerate(board) if cell is None]
        if not available:
            raise ValueError("No moves available")
        if difficulty == "easy":
            return random.choice(available)
        if difficulty == "medium":
            win = self._find_winning_move(board, "O")
            if win is not None:
                return win
            block = self._find_winning_move(board, "X")
            if block is not None:
                return block
            if 4 in available:
                return 4
            corners = [c for c in [0, 2, 6, 8] if c in available]
            if corners:
                return random.choice(corners)
            return random.choice(available)
        return self._get_best_move(board)

    def _find_winning_move(
        self, board: List[Optional[str]], player: str
    ) -> Optional[int]:
        for line in self.WINNING_LINES:
            vals = [board[i] for i in line]
            if vals.count(player) == 2 and vals.count(None) == 1:
                return next(line[i] for i in range(3) if board[line[i]] is None)
        return None

    def _get_best_move(self, board: List[Optional[str]]) -> int:
        best_score = -float("inf")
        best_move = 0
        for i in range(9):
            if board[i] is None:
                board[i] = "O"
                score = self._minimax(board, 0, False)
                board[i] = None
                if score > best_score:
                    best_score = score
                    best_move = i
        return best_move

    def _minimax(
        self, board: List[Optional[str]], depth: int, is_maximizing: bool
    ) -> int:
        winner = self._check_winner(board)
        if winner == "O":
            return 1
        if winner == "X":
            return -1
        if self._is_board_full(board):
            return 0
        if is_maximizing:
            best = -float("inf")
            for i in range(9):
                if board[i] is None:
                    board[i] = "O"
                    best = max(best, self._minimax(board, depth + 1, False))
                    board[i] = None
            return best
        best = float("inf")
        for i in range(9):
            if board[i] is None:
                board[i] = "X"
                best = min(best, self._minimax(board, depth + 1, True))
                board[i] = None
        return best

    def _check_winner(self, board: List[Optional[str]]) -> Optional[str]:
        for line in self.WINNING_LINES:
            if board[line[0]] and board[line[0]] == board[line[1]] == board[line[2]]:
                return board[line[0]]
        return None

    def _is_board_full(self, board: List[Optional[str]]) -> bool:
        return None not in board


tic_tac_toe_game = TicTacToe()
